state/state.go

```diff
@@ -2168,6 +2168,19 @@ func (st *State) AddRelation(eps ...Endpoint) (r *Relation, err error) {
 			if err != nil {
 				return nil, err
 			}
+			// Enforce max-relation limits for the app:ep combination
+			existingRelationCount, err := enforceMaxRelationLimit(app, ep)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			assertions := bson.D{{Name: "life", Value: Alive}}
+			if existingRelationCount >= 0 {
+				relations, err := app.Relations()
```

Comment:
Have `enforceMaxRelationLimit` return this value instead of doing a second read. Because, you are making an txn assertion about the second read but this is all predicated on the first read of relations.

Reply:
done